### PR TITLE
feat: use gen_random_uuid() for CockroachDB INSERTs

### PR DESCRIPTION
**Pull request by alnr**
Using the `gen_random_uuid()` builtin can significantly speed up inserts because the primary key uniqueness constraint check is elided.

I've also fixed up code where there was a prepared statement being created for just one query, which is super inefficient.

I'm kinda having trouble running the tests locally, so I'd appreciate a workflow approval too. @aeneasr @sio4 👍

### CODE REVIEW

**Comment by aeneasr** [comment]
Approval given :)

You can run the tests locally using `./test.sh` which executes tests against all DBs, or using:

```
go build -v -tags sqlite -o tsoda ./soda

SODA_DIALECT=cockroach

./tsoda drop -e $SODA_DIALECT -c ./database.yml -p ./testdata/migrations
./tsoda create -e $SODA_DIALECT -c ./database.yml -p ./testdata/migrations
./tsoda migrate -e $SODA_DIALECT -c ./database.yml -p ./testdata/migrations

go test -cover -race -tags sqlite -count=1 $args ./...
```

File: dialect_cockroach.go [hunk 1]
```diff
@@ -77,27 +78,55 @@ func (p *cockroach) Create(c *Connection, model *Model, cols columns.Columns) er
...
 		if len(w.Cols) > 0 {
-			query = fmt.Sprintf("INSERT INTO %s (%s) VALUES (%s) returning %s", p.Quote(model.TableName()), w.QuotedString(p), w.SymbolizedString(), model.IDField())
+			query = fmt.Sprintf("INSERT INTO %s (%s) VALUES (%s) RETURNING %s", p.Quote(model.TableName()), w.QuotedString(p), w.SymbolizedString(), model.IDField())
 		} else {
-			query = fmt.Sprintf("INSERT INTO %s DEFAULT VALUES returning %s", p.Quote(model.TableName()), model.IDField())
+			query = fmt.Sprintf("INSERT INTO %s DEFAULT VALUES RETURNING %s", p.Quote(model.TableName()), model.IDField())
 		}
 		txlog(logging.SQL, c, query, model.Value)
-		stmt, err := c.Store.PrepareNamed(query)
+		rows, err := c.Store.NamedQueryContext(model.ctx, query, model.Value)
```
Comment (aeneasr) [inline dialect_cockroach.go:86]: Good point, it makes no sense to use prepared statements if they're not cached.

**Review by aeneasr — COMMENTED** [review]
This looks very nice already! Do we have a test showing that the ID is generated correctly for Cockroach? I do believe that such tests exist, but it would be good to have proof that we don't introduce regressions :)

**Comment by alnr** [comment]
- A couple of tests were not idempotent. I think I've fixed them all.
- Fixed missing Close on sql.Rows
- Added more explicit checks about UUID generation to test

ready for review :)

File: database.yml [hunk 2]
```diff
@@ -1,25 +1,27 @@
...
-  user: {{ envOr "MYSQL_USER"  "root"  }}
-  password: {{ envOr "MYSQL_PASSWORD"  "root#"  }}
+  host: '{{ envOr "MYSQL_HOST" "127.0.0.1"  }}'
+  port: '{{ envOr "MYSQL_PORT" "3306"  }}'
+  user: '{{ envOr "MYSQL_USER"  "root"  }}'
+  password: '{{ envOr "MYSQL_PASSWORD"  "root#"  }}'
   options:
     readTimeout: 5s
+    parseTime: true
+    multiStatements: true
```
Comment (sio4) [inline database.yml:11]: what is the reason for this change? (single quotes and adding options) I cannot find any description in the PR content. please revert them all if they are not related to the topic of this PR.

File: docker-compose.yml [hunk 3]
```diff
@@ -34,7 +35,7 @@ services:
       start_period: 3s
 
   cockroach:
-    image: cockroachdb/cockroach:latest-v21.1
+    image: cockroachdb/cockroach:latest-v22.2
```
Comment (sio4) [inline docker-compose.yml:38]: Don't do this too. I don't think the minimum supported version is a part of this PR. If there is any issue with versions, please explain the reason in the PR description.

**Review by sio4 — CHANGES_REQUESTED** [review]
Hi @alnr,
Thank you for your PR.

Before reviewing the main topic of this PR, please remove changes regarding testing configurations (database.yml and docker-compose.yml). I don't think they are parts of the topic of this PR. Changing the supported database version should have more consideration around other conditions and may not be a part of this feature PR.

Please keep PR as simple as possible with a single point of interest.

File: database.yml (hunk 2 above)
Comment (alnr) [reply database.yml]: The pop migrator complained about this in the logs.

Reverted.

File: docker-compose.yml (hunk 3 above)
Comment (alnr) [reply docker-compose.yml]: Reverted.

**Comment by sio4** [comment]
Hi @alnr,

I am just wondering. What could be the cons and pros of using `gen_random_uuid()` for generating a random UUID ID on the server side?

Personally, I prefer to use application-side actions (spending the application server's CPU) rather than offloading CPU tasks to the database side since the application is more flexible to spread out (scaling) than controlling database load unless the action should be taken on the database side for integrity or similar things. Also for the main topic of this PR, it is easier to maintain the code if we use the same generic method in the generic dialect.

Could you please explain your idea and/or considerations?

**Comment by alnr** [comment]
> I am just wondering. What could be the cons and pros of using `gen_random_uuid()` for generating a random UUID ID on the server side?

Pros:
- primary key uniqueness constraint check will be elided on insert. This makes a huge difference in CockroachDB multiregion databases. If you don't use `gen_random_uuid()`, the uniqueness constraint will be verified prior to insert on all regional shards, which requires contacting them. Across continents, that's multiple hundred milliseconds. I've confirmed this with a US-EU-APAC cluster on a hosted enterprise Cockroach cluster.

Cons:
- can't think of any. Generating a UUID is free. Doesn't matter if the app or the DB does it.
 
> Personally, I prefer to use application-side actions (spending the application server's CPU) rather than offloading CPU tasks to the database side since the application is more flexible to spread out (scaling) than controlling database load unless the action should be taken on the database side for integrity or similar things. Also for the main topic of this PR, it is easier to maintain the code if we use the same generic method in the generic dialect.

After this change, we burn a lot fewer CPU cycles anyway, because the DB does not have to verify the uniqueness constraint on the primary key. Checking that constraint takes approximately infinity times more time than generating a UUID.

**Comment by alnr** [comment]
Not sure what's going on with that failing test 🤷

**Comment by sio4** [comment]
> Not sure what's going on with that failing test shrug

Don't worry. The short test on the Windows machine is failing many times, and I don't think it is related to your PR. We need to make it better but actually, the Windows environment is too hard. :-)

For the main topic, I quickly researched the Cockroach's resources and I found the followings:

* https://www.cockroachlabs.com/blog/regional-by-row/#enforcing-uniqueness-with-implicit-partitioning
* https://www.cockroachlabs.com/docs/stable/schema-design-table.html#primary-key-examples
* https://github.com/cockroachdb/cockroach/pull/61132
* https://github.com/cockroachdb/docs/issues/10118

It seems like we can take another approach for this, not for right now but for later:

* improve the generic routine can allow empty UUID id value (with a new option)
* improve fizz to create a better `CREATE` statement with the default `gen_random_uuid()` (or manually?)

I think this approach will make Pop simple without losing the database side performance. Any ideas or suggestions will help us greatly!

> Checking that constraint takes approximately infinity times more time than generating a UUID.

Yeah, could be in a global cluster _across the universe_ :laughing:

**Review by sio4 — APPROVED** [review]

**Comment by alnr** [comment]
Thanks for merging!

Honestly I would keep the code as-is for now. Not everyone uses fizz to create schemas. Regarding the empty UUID value: I don't think allowing null UUIDs makes sense *for the ID (primary key) column*. It makes a lot of sense for other columns, of course, but that's not where this optimization is most useful anyway.

**Comment by sio4** [comment]
> Honestly I would keep the code as-is for now. Not everyone uses fizz to create schemas. Regarding the empty UUID value: I don't think allowing null UUIDs makes sense _for the ID (primary key) column_.

Yes, for now. and...
Maybe my explanation is not enough. I don't mean `ID` can be null UUID. Currently, we remove an integer `ID` field from the model when executing `Create()` because we assume auto-increment is enabled on the database side. With the same concept, if we know (with option or anything) the database table was created with a UUID key with `DEFAULT` like below (example taken from the cockroach's document), we can just remove the ID field so the database will handle it.

```sql
CREATE TABLE movr.max_schema.vehicles (
      id UUID DEFAULT gen_random_uuid() PRIMARY KEY,
      type movr.max_schema.vtype,
      creation_time TIMESTAMPTZ,
      available BOOL,
      last_location STRING
  );
```

The function or method to create an auto key can be different by database engines, but we can make the `Create()` routine the same.

There are also related issues (opposite actually) for the integer ID that cannot be preserved when if the user sets `ID` explicitly and I think those issues can be considered together with it. (#382, #586)
